cilium/cmd/debuginfo.go

```diff
@@ -77,59 +96,172 @@ var sections = map[string]addSection{
+	fmt.Fprint(os.Stderr, "Warning, some of the BPF commands might fail when not run as root\n")
+}
+
+func runDebugInfo(cmd *cobra.Command, args []string) {
+
+	validateInput()
 
 	resp, err := client.Daemon.GetDebuginfo(nil)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "%s\n", pkg.Hint(err))
 		os.Exit(1)
 	}
 
-	// define output type and file path
 	var output outputType
-	var path string
-
-	switch {
-	case len(file) > 0: // Markdown file
-		output = MARKDOWN
-		path = file
-	case len(html) > 0: // HTML file
-		output = HTML
-		path = html
-	default: // Write to standard output
-		output = STDOUT
-	}
 
 	// create tab-writer to fill buffer
 	var buf bytes.Buffer
 	w := tabwriter.NewWriter(&buf, 5, 0, 3, ' ', 0)
 	p := resp.Payload
 
-	// generate multiple files
-	if (len(file) > 0 || len(html) > 0) && filePerCommand {
-		for cmdName, section := range sections {
+	cmdTime := time.Now()
+
+	if outputToFile && len(outputOpts) == 0 {
+		outputOpts = append(outputOpts, markdownOutput)
+	}
+
+	// Dump payload for each output format.
+	for _, outputOpt := range outputOpts {
+		var fileName string
+
+		switch strings.ToLower(outputOpt) {
+		case markdownOutput:
+			output = MARKDOWN
+		case htmlOutput:
+			output = HTML
+		case jsonOutputDebuginfo:
+			output = JSONOUTPUT
+		case jsonpathOutput:
+			output = JSONPATH
+		default:
+			if jsonPathRegExp.MatchString(outputOpt) {
+				output = JSONPATH
+			}
+		}
+		// Only warn when not dumping output as JSON so that when the output of the
+		// command is specified to be JSON, the only outputted content is the JSON
+		// model of debuginfo.
+		if os.Getuid() != 0 && output != JSONOUTPUT && output != JSONPATH {
+			rootWarningMessage()
+		}
+
+		if outputToFile {
+			fileName = formatFileName(outputDir, cmdTime, output)
+		}
+
+		// Generate multiple files for each subsection of the command if
+		// specified, except in the JSON cases, because in the JSON cases,
+		// we want to dump the entire DebugInfo JSON object, not sections of it.
+		if filePerCommand && (output != JSONOUTPUT && output != JSONPATH) {
+			for cmdName, section := range sections {
+				addHeader(w)
+				section(w, p)
+				writeToOutput(buf, output, fileName, cmdName)
+				buf.Reset()
+			}
+			continue
+		}
+
+		// Generate a single file, except not for JSON; no formatting is
+		// needed.
+		if output == JSONOUTPUT || output == JSONPATH {
+			marshaledDebugInfo, _ := p.MarshalBinary()
+			buf.Write(marshaledDebugInfo)
+			if output == JSONOUTPUT {
+				writeToOutput(buf, output, fileName, "")
+			} else {
+				writeJSONPathToOutput(buf, fileName, "", outputOpt)
+			}
+			buf.Reset()
+		} else {
 			addHeader(w)
-			section(w, p)
-			writeToOutput(buf, output, path, cmdName)
+			for _, section := range sections {
+				section(w, p)
+			}
+			writeToOutput(buf, output, fileName, "")
 			buf.Reset()
 		}
+	}
+
+	if len(outputOpts) > 0 {
 		return
 	}
 
-	// generate a single file
+	if os.Getuid() != 0 {
+		rootWarningMessage()
```

Comment:
This is added because if the user specifies no output type, we don't range over the list of output types. The default output type is markdown, so we don't need to provide the JSON check.